### PR TITLE
[Fix] Patched RCE arising inside the "launchpad" library

### PR DESCRIPTION
**Pull request by Mik317**
### 📊 Metadata *

_Please enter the direct URL for this bounty on huntr.dev. This is compulsory and will help us process your bounty submission quicker._

### ⚙️ Description *

The issue arised in multiple locations, so I to validate the type of data the functions were going to use (like the `paths`) and since there were some `multi commands` I didn't use the `execFile` function, which should have had stored many new variables because we wouldn't have been to concatenate and return results that would have been used in a second/third command correlated to the first one executed. In this case I simply made a functionality that `deletes every quote` from the variable, making impossible threat the `variables concatenated` as commands, but only as `arguments` of the specific command.

### 💻 Technical Description *

The fix has been applied in 3 different ways inside 3 different files, so I'll comment each one.

1. The issue arises firstly here: https://github.com/bitovi/launchpad/blob/master/lib/local/instance.js#L12 because of the fact the `name` variable is `concatenated` inside the various commands without being sanitized. Since the `name` is inside some `single-quotes` it would have been useless split the 3 different commands inside 3 different `execFile` that would have used more resources to store the content of the singular commands that should be concatenated again ... instead I introduced the `safe` function which deletes the `quotes` from the `name` in order to make it to be only an argument not escapable from quotes. 

```js
var safe = function (str) {
   // Avoid quotes makes impossible escape the `multi command` scenario
   return str.replace(/['"]+/g, '');
}
```
Note I've used the `execFile` function later in this file in the following lines: https://github.com/Mik317/launchpad/blob/master/lib/local/instance.js#L104 and https://github.com/Mik317/launchpad/blob/master/lib/local/instance.js#L110, in order to avoid `commands` could be executed in a dangerous context.
2. The 2' issue arised inside the following line: https://github.com/bitovi/launchpad/blob/master/lib/local/version.js#L21
In this case it I used `execFile` in order to avoid concatenation of other strings containing dangerous characters.
Patched with:

```js
    exec(command.split(' ')[0], command.split(' ').slice(1), function(error, stdout) {
```
In this case the first part of the `command` is taken as `command to execute` (surely a path since the `command` variable is made through 

```js
    var command = path.join('"' + __dirname, '..', '..', 'resources', 'ShowVer.exe" "' + browser.command + '"');
```
), while the second part are the `arguments`.

3. The last issue arised here: https://github.com/bitovi/launchpad/blob/master/lib/local/version.js#L50
In this case the `browser path and filename` weren't checked completely, and even if the execution of malicious code would have been possible only if the `default browser` of the victim has a badly crafted `filename`, I inserted a check to see if the `path+filename` pointing to the browser is a valid `path`.
The issue has been fixed through this function:

```js
// Validate paths supplied by the user in order to avoid "arbitrary command execution"
var validPath = function (filename){
  var filter = /[`!@#$%^&*()_+\-=\[\]{};':"\\|,<>\/?~]/;
  if (filter.test(filename)){
    console.log('\nInvalid characters inside the path to the browser\n');
    return
  }
  return filename;
}
```

### 🐛 Proof of Concept (PoC) *

1. Download the JS library (launchpad)
2. Go inside the path you downloaded it and make the following `poc.js` file:

```js
// poc.js
var launchpad = require("./launchpad/lib/local/instance");
var tst = new launchpad.Instance('t', {}, {}, {process:"s'; touch HACKED; # "});
tst.getPid(function(){});
```
3. Execute through `node poc.js`
4. The `HACKED` file will we created
Screenshot from 2020-06-26 19-34-24

### 🔥 Proof of Fix (PoF) *

1. Download the fixed version
2. Use the same POC previously indexed
3. The `HACKED` file is NOT created anymore 
Screenshot from 2020-06-26 19-52-10

### 👍 User Acceptance Testing (UAT)

It doesn't introduce any error (at least using the module through the PoC I crafted)

Regards, 
Mik

### CODE REVIEW

File: lib/local/instance.js [hunk 1]
```diff
@@ -90,11 +97,11 @@ Instance.prototype.stop = function (callback) {
     } catch (error) {}
   } else {
     if (this.options.command.indexOf('open') === 0) {
-      command = 'osascript -e \'tell application "' + self.options.process + '" to quit\'';
+      command = 'osascript -e \'tell application "' + safe(self.options.process) + '" to quit\'';
       debug('Executing shutdown AppleScript', command);
       exec(command);
```
Comment (toufik-airane) [inline lib/local/instance.js]: I think that the dangerous function `exec()` should be avoided.

**Review by toufik-airane — CHANGES_REQUESTED** [review]

File: lib/local/instance.js (hunk 1 above)
Comment (Mik317) [reply lib/local/instance.js]: Hi @toufik-airane :),
I've replaced the `exec` function with the `execFile` one in order to avoid problems. I removed also the `safe` function in the `windows` case because I saw certain systems won't recognize the string inside as a argument (edge cases covered anyway by `execFile` now introduced).

Regards,
Mik

Comment (toufik-airane) [reply lib/local/instance.js]: Have you run the tests and update the description?

Comment (Mik317) [reply lib/local/instance.js]: Yeah, I updated the description of the `fixes` applied and I retested the new code.
I saw I had missed a `)` ... however now it should be fixed well and running the POC I previously used it doesn't lead to `RCE` or errors due to malformed syntax.

Regards,
Mik

Comment (toufik-airane) [reply lib/local/instance.js]: Thank you. 😊

**Review by toufik-airane — APPROVED** [review]

**Review by toufik-airane — APPROVED** [review]

File: lib/local/version.js [hunk 2]
```diff
@@ -18,7 +28,7 @@ module.exports = function(browser) {
 
     debug('Retrieving version for windows executable', command);
     // Can't use Q.nfcall here unfortunately because of non 0 exit code
-    exec(command, function(error, stdout) {
+    exec(command.split(' ')[0], command.split(' ').slice(1), function(error, stdout) {
```
Comment (mufeedvh) [inline lib/local/version.js]: Can you please change this `exec()` to `execFile()`

**Review by mufeedvh — COMMENTED** [review]
Good fix, just complete this one change request and it's all perfect! :clap::tada:

![huntr sheriff](https://img.shields.io/badge/SHERIFF-red?labelColor=black&link=https://www.huntr.dev&style=flat-square&logo=data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAfQAAAH0BAMAAAA5+MK5AAAACXBIWXMAAAsTAAALEwEAmpwYAAAAFVBMVEVHcEzUTjrUSz7bUUXfTj/eTD3wWkwE37jxAAAABnRSTlMAOsPzjWJ73M2lAAAJoElEQVR42u2dS0PbOhBGLb/W4eU1hdbrtIDXAVKv01K8DjHo//+EC5RbHNsQSbHs0eg76yblRLI0I43kIAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAECbg4OZl97H5/KZzU/v7JNcvlFf+2V+KBtc+WR+J7d49NbcI/dD2cGTPh/LHr74YC6KPvV65YF6JXt59LS7+9Hli4/Ua+5h3ZH8kCcfx7g3eI902SfmvEc6IT9l4Wujs272HY3O+Wk/2qXOd5AvpK/NHu00lw9M1fPd6kxDulgqcOpTyubD/CaV4DjQhWrqHAe6Uk1942EkxzeQj1TV+fX4XFW95maeSOlrj4/U1bn1+FJdnVmPF1L62uMjHXVePb7SUecV1UgtOMXxsZ76nJF6pjG+f/tzzmmJLtdo8noVCE+nNrnhtEgV6j3qnPp7pafOaYEu11TnM7EnUpe5p486p2bPtNXZPO25vjqT5XghDeARxscm6jwGutREnUdYU5qo8xjjCxN1ySGQT4zMWSzQhWbqaz8DGi7jXGmmvvF2lOMwzglDcwbjXGyq7n48l5qqu78H87o4VUsfk7eXAf7SZJR3f8P1WeJ76OUQL15OLuZeDvHh8zMb+jnER5uZYaM7H8VnC+OZ3fUh/l5/74VPFG8cyUrn1SNjdecTmFz6msB8sD51cbZcLu/vl8ubm/OC6+zWVfrx81drOPhzUzCc3Tr9/Wv/I/z3hh5OuZtQE3+VL3hN7Nv9vf716T++ZTWxV1oFQoecJvatDrx7oj7kox7rFoUdsYlpUu0t85JLTFNqX76SMCkrEQbD9REP9dhAQ/AI50KTOTpjoR6ZpCIJC/XUaJ7KOQTxqVE4nnHYY8+Mum7MSl0vOOGQumVm0XjOSF0z/awYZK2V2UObMmj1ymyCDhm1+lzvYzGjVtfMQxIGrZ6ZqQtG6gszdQ6T2wrqmuGcy4Fsul+re6geMEhaw/2GuTXUXSTeb16f+6eeMFqHN4zhFwzUDTO3lcPqgZl6xGG7MTeaoFMOm8ylUTCecaiIr/ZZm3O7oCTbZ0XW7VN+ZgMWi+Ixo0hWsCgZjE1imphBMPfegA8mT8nCafVgn53WmdvqucEUXfI4ClAZtCCTAyAGK/EJk2M/kf44FzEphzfYPquYHIIQ+nVEBZez+9olgwm7A08Puo+6+4e4tYtDSjZH91PNa9QEnwsbQk2VkM81HUKzQLhicrbx5LpxJc9K65d6HuCX7o7xJ+cvTV1q9fjo/ZaO52jocuWkeHL7t6kzrStHyvf62NcPuij/+/+mDnUOQsSNWvi3iOCbY90+Lv4NbonOQc33E4Gn/z5XX7tkftc84lWoH95qnHuZNe6pu3Sm4UW+dRy5VG/2rHFWpvklm5Vjnf1tcMuUz+w1Gv2pda+JE53+uH39c6h8LLtx9nfdvpfxG33zw871MkL1vGLzYON15yD/FXXznhfwrnLFI73N46w933PlnLl8qtQuYWj2l03f3RVXTvX2V34r3biS7L5/8zvhsf2DG5WU3r+rcn/NF6rmCu32cfJ6p3TzINVMVvHeoUdjc7Kvw1H863vdlT/75NCD3uvebrtbty+m0nqF1VZcmuh8lOAm5JHerWKX/y7nEbd6nyQ3wwntW7E3X+8PDu6X59r3y1Ib6QxvgWdwo7J+o5tDrNkjOSKnzg7ve0NqkI/lqKz8HOSoDXTFuOqE9p8TOTIzP8d3WmN8Nbb6g59TG62HfWxzOiVGo49ydCqH4/HVV74O8HSGeI/V0/HV535G8JQKxtvql/e/Dgbl5OSGqHo1wk74Mc1wrmxHWgcHs5nd35fKTkTe2iN47qLDp1aCZCRbbP9RQvw5+2J5RKGyRtWedsKLix/LJjdnZ2dfl3vwsx0z1hTVV4G4qy+GXra5av83RNbgW82RWAraS3rqSeshtBLXPrTGeBpLVHFr1rES3D22vpdG6ha21EtL2w4pvW32SdRppG5RS70aQ51G6pZ+9kzaetZp5C8Z1EdVp5G6Va1n0op6e9KkkbqVnw1HA6pX9FK3fAz1h27vIqhuK5At6anLKdQlPfXakvqaorpoq4eW1HNy1QVJeyNwHHUKqVs8jfqCorqVjdd5W51C6haOpV6Q23WL2sVN/qino6iftstWKKRuWVs98VV9NZI6hdStaqsLO+qCnnprFXJhR73ztY/+qlNI3fJR1DvPEYUNx13j0VDqCb2qQX/VxTjq3XBh+qw16aoXVtRjcrWiya5gexi6WdH0WWu8K7EeSj0kVysaTqU+feoWTaU+ff6SQv3DVeNB6FnonX7DMRtLPZLUUrdq1zbJUEUVkaSWulU79gXt1JOQSN3KqdSnT93yHRvhgxVVpJJa/lLsKH+wU09CQl3lj7SjPnXqJsZR7+tLK2rqm9HUp07dEoWx2EIVEYXULR5HvW/KnDp1CxXiLhulNATUI4Vo25L62g/1vqRo6tQtVUgvbdSTEEjdMoVFBUvqU6duFdQtq8971KfOWkuFFXMbpTQEstZcYZ/Ekrr0Q713J4+geuKHeqmwHWxJvSaoLmyoC3Lq1e6qFyulNAQmt8xf9VSpbw5fRUQgmovUBuPhS2mmT1/C6dSnzteTcdT7goWpV2mEWsw5fBURgYqSYhT1vpxo6nX43oXS3IJ6SC2O7Znd1iOpT7/JHE6lPn1VhRhDve7/ialtsK8tVJT0rXAvplfPJlKf3rwz69ioIyJZRdSd2W0U0/TsYZK4p6OyX0zTo74I6PV4G8U03XoFInft5dbVu1UqRK4UPbJeTNP9zgUNdTG+OplXv2S2K0o6Q+eainoyuvosoNjsNtTbsQKhd940X+xlo46oHSHSafStvKq2r07rRX6lVfXtRJjYG0vfu7yNOqJt9dOAFrHNipItdXqvaj0aSZ3iC3rv7Kk3V3lXAUFKa3VEjbX9RUCSO1vFNO/q1wFRfltS/7ePR9Y8CI6Llw6fWFKvVwFhxO2VHfXn3/RyFpBHWFC/q6+DwEP1ReAKUB+ynsQZoD5gPQnU6VNAfbiiCmfIoQ51qO9TT+IMA1eUbBxSH3ir9dEh9YHXJeeBrw97PXNJPTm/GI5VAAAAFDm5H4KZg+YDRXQb99wHi2q+O6de+RjFDh3PQR3qUPditWLjnPpgZYMPzqmLodbiHUxZk7NBuA4AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAIAU/wGK2a+A3NJU1QAAAABJRU5ErkJggg==)

File: lib/local/version.js (hunk 2 above)
Comment (JamieSlome) [reply lib/local/version.js]: @Mik317 - once you have made the fix - we will get this merged - cheers! 🍰

Comment (Mik317) [reply lib/local/version.js]: Ok, I just updated the fix as requested. In the meanwhile I also applied a `minor fix` because I had inserted certain characters as `blacklisted` that however could be part of a `path` (not only of a `filename`) :).

Regards,
Mik

Comment (JamieSlome) [reply lib/local/version.js]: @Mik317 - awesome, thanks for making the changes!

**Comment by huntr-helper** [comment]
Congratulations Mik317 - your fix has been selected! 🎉

Thanks for being part of the community & helping secure the world's open source code.
If you have any questions, please respond in the comments section. Your bounty is on its way - keep hunting!
